File: solutions/49.group-anagrams.java
/*
 * @lc app=leetcode id=49 lang=java
 *
 * [49] Group Anagrams
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<Map<Character, Integer>, List<String>> res = new HashMap<>();
        for (String s : strs) {
            Map<Character, Integer> key = new HashMap<>();
            for (Character c : s.toCharArray()) {
                key.computeIfAbsent(c, k -> 0);
                key.computeIfPresent(c, (k, value) -> value + 1);
            }
            res.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        return new ArrayList<>(res.values());
    }

    public List<List<String>> groupAnagramsV1(String[] strs) {
        Map<String, List<String>> res = new HashMap<>();
        for (String s : strs) {
            char[] c = s.toCharArray();
            Arrays.sort(c);
            String key = new String(c);
            res.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        return new ArrayList<>(res.values());
    }
}
// @lc code=end

File: solutions/49.group-anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#

# @lc code=start
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        resMap = defaultdict(list)
        for s in strs:
            key = [0] * 26
            for c in s:
                key[ord(c) - ord("a")] += 1
            resMap[tuple(key)].append(s)

        return list(resMap.values())


    def groupAnagramsV1(self, strs: List[str]) -> List[List[str]]:
        resMap = defaultdict(list)
        for s in strs:
            key = tuple(sorted(s))
            resMap[key].append(s)

        return list(resMap.values())

# @lc code=end

